### ISSUE TITLE
add svg caching feature 

### ISSUE THREAD

**Issue by HosamHasanRamadan** [zathras/jovial_svg#6]
when I fetch SVG from the internet using `ScalableImageSource.fromSvgHttpUrl` in `ListView.builder` every list item rebuild will re-fetch the svg again.
Thank you for the package it is really faster and better than `flutter_svg`  package

**Comment by zathras** [comment]
Hi Hasan,

I guess that this is an enhancement request, suggesting that a caching feature would be worthwhile?  Sound reasonable.

Would it be possible for you to put together a small example of the usage you have in mind, and attach the .dart file(s)?  That way I can be confident I'm addressing the right issue.  Also, please confirm it's OK to include your code in the demo app.

**Comment by HosamHasanRamadan** [comment]
for the sake of simplicity i generated 100 identical SVG URL  
```dart
import 'package:flutter/material.dart';
import 'package:jovial_svg/jovial_svg.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: HomePage(),
    );
  }
}

class HomePage extends StatelessWidget {
  final svgs = List.generate(
      100,
      (index) => Uri.parse(
          'https://raw.githubusercontent.com/feathericons/feather/master/icons/activity.svg'));

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('jovial_svg_issue'),
      ),
      body: GridView.builder(
          itemCount: svgs.length,
          gridDelegate:
              SliverGridDelegateWithFixedCrossAxisCount(crossAxisCount: 5),
          itemBuilder: (context, index) {
            return GridTile(
              child: ScalableImageWidget.fromSISource(
                  si: ScalableImageSource.fromSvgHttpUrl(svgs[index])),
            );
          }),
    );
  }
}

```

**Comment by zathras** [comment]
ش كرًا.

Note to self:  githubusercontent strips parameters, so I can make the URLs distinct by adding one, viz:  https://raw.githubusercontent.com/feathericons/feather/master/icons/activity.svg?x=1

**Comment by zathras** [comment]
Hi Hosam,

Does `ScalableImageCache` look reasonable, and address what you're looking for?

Also, I'm wondering how soon I should spin a new release.  It's not a big deal or anything, but if you think it's likely you'll find more issues in the next week or two, I might hold off going through the release process.

Cheers,

Bill

**Comment by HosamHasanRamadan** [comment]
I used the main branch right now but it seems buggy :
```shell
'package:jovial_svg/src/widget.dart': Failed assertion: line 680 pos 14: 'e._refCount > 0': is not true.
```

![error image](https://imgur.com/aChIQQN.png)

**Comment by zathras** [comment]
Hmmm...  This has me puzzled.  Could you bring over the latest, and re-run?  It's doing more consistency checks now.

This is maybe a long shot, but did you by any chance provide your own subtype of `ScalableImageSource`?  Are you sure `hashCode` is correctly implemented?

Failing that, and assuming you still get an error, could you give me some more context?  Like, how big is the cache, what kind of ScalableImageSource is in it, how hard is it to reproduce, and stuff like that.  If you could make a (small-ish) self-contained example that reproduces the problem reliably, that would of course be ideal.

The following updates do some more checking, including guarding against a bad `hashCode` implementation:
https://github.com/zathras/jovial_svg/commit/81ee71da10594d90b5a47edfdaf4215650178553
https://github.com/zathras/jovial_svg/commit/5072f607231dfa45c8dd1e89947daca2648b34d8

**Comment by zathras** [comment]
Hi Hosam,

Check out the latest commit.  I wasn't handling the case of different caches when Flutter recycles my widget's state object - that would explain the exception, too.  I bet this was the problem.

If not, please give a full stack trace.

**Comment by zathras** [comment]
Good, I'm glad it was just "flutter clean."

Interesting about hive...  See https://github.com/zathras/jovial_svg/issues/12

**Comment by zathras** [comment]
Published in 1.0.7
